service/matching/version_rule_helpers.go

```diff
@@ -69,8 +69,16 @@ var (
 	errExceedsMaxAssignmentRules = func(cnt, max int) error {
 		return serviceerror.NewFailedPrecondition(fmt.Sprintf("update exceeds number of assignment rules permitted in namespace (%v/%v)", cnt, max))
 	}
-	errRequireUnconditionalAssignmentRule = serviceerror.NewFailedPrecondition("there must exist at least one fully-ramped 'unconditional' assignment rule, use force=true to bypass this requirement")
-	errExceedsMaxRedirectRules            = func(cnt, max int) error {
+	// errRequireFullyRampedAssignmentRule is thrown if the task queue previously had a fully-ramped assignment rule and
+	// the requested operation would result in a list of assignment rules without a fully-ramped assigment rule, which
+	// effectively means that the task queue does not have a default version. This error is only thrown when moving from
+	// a task queue with a fully-ramped assignment rule (aka versioned with a default Build ID / default version) to a
+	// task queue without any fully-ramped assignment rules, because that operation would make the unversioned queue the
+	// default queue. It's important that users do not accidentally make the unversioned queue the default after the
+	// unversioned worker has already been decommissioned, so we throw this error. If you are intentionally reverting from
+	// a versioned default Build ID to an unversioned default Build ID, use force=true to bypass this requirement.
+	errRequireFullyRampedAssignmentRule = serviceerror.NewFailedPrecondition("there must exist at least one fully-ramped assignment rule, use force=true to bypass this requirement and make the unversioned queue the default")
```

Comment:
please review this comment since I'm not sure I'm doing a good enough job here

Reply:
(this replaces `errRequireUnconditionalAssignmentRule` above)

Reply:
lgtm

Reply:
I don't understand lowercased developer-facing errors, whether emitted from preconditions or assertions, or in logging. That said, "At least one fully-ramped assignment rule must exist. Use force=true to bypass this requirement and set the unversioned queue as the default."

Reply:
I like that better, thanks. I think the lower case convention comes from wanting to be able to compose multiple errors separated by colons without having capitals in the middle of the composed error string, but I'm not sure